core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+// consumption pattern.
+//
+// BankingSimulator consists of 2 phases chronologically: warm-up and on-the-fly. The 2 phases are
+// segregated by the aforementioned T=0.
+//
+// Both phases just sends BankingPacketBatch in the same fashion, pretending to be sigveirfy
+// stage/gossip while burning 1 thread to busy loop for precise T=N at ~1us granularity.
+//
+// Warm up is defined as T=-N secs using slot distance between immediate ancestor of first
+// simulated block and root block. As soon as warm up is initiated, we invoke
+// `BankingStage::new_num_threads()` as well to simulate the pre-leader slot's tx-buffering time.
+pub struct BankingSimulator {
+    banking_trace_events: BankingTraceEvents,
+    first_simulated_slot: Slot,
+}
+
+#[derive(Error, Debug)]
+pub enum SimulateError {
+    #[error("IO Error: {0}")]
+    IoError(#[from] io::Error),
+
+    #[error("Deserialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+}
+
+pub struct BankingTraceEvents {
+    packet_batches_by_time: BTreeMap<SystemTime, (ChannelLabel, BankingPacketBatch)>,
+    freeze_time_by_slot: BTreeMap<Slot, SystemTime>,
+    hash_overrides: HashOverrides,
+}
+
+impl BankingTraceEvents {
+    fn read_event_file(
+        events: &mut Vec<TimedTracedEvent>,
+        event_file_path: &PathBuf,
+    ) -> Result<(), SimulateError> {
+        let mut reader = BufReader::new(File::open(event_file_path)?);
+
+        loop {
+            events.push(deserialize_from(&mut reader)?);
+
+            if reader.fill_buf()?.is_empty() {
+                // EOF is reached at a correct deserialization boundary.
+                // We're looking-ahead the buf, so NOT calling reader.consume(..) is correct.
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn load(event_file_paths: Vec<PathBuf>) -> Result<Self, SimulateError> {
+        let mut events = vec![];
+        for event_file_path in &event_file_paths {
+            let old_len = events.len();
+            let _ = Self::read_event_file(&mut events, event_file_path).inspect_err(|error| {
+                error!(
+                    "Reading {event_file_path:?} failed after {} events: {:?} due to file corruption or unclearn validator shutdown",
+                    events.len() - old_len,
+                    error
+                );
+            });
+            info!(
+                "Read {} events from {:?}",
+                events.len() - old_len,
+                event_file_path
+            );
+        }
+
+        let mut packet_batches_by_time = BTreeMap::new();
+        let mut freeze_time_by_slot = BTreeMap::new();
+        let mut hash_overrides = HashOverrides::default();
+        for TimedTracedEvent(event_time, event) in events {
+            match event {
+                TracedEvent::PacketBatch(label, batch) => {
+                    // Somewhat naively assume that event_times (nanosecond resolution) won't
```

Comment:
with our current system perf we've definitely not got to worry about that for a while 😆